File: 2024/java/day6_2.java
import java.io.File;
import java.io.FileNotFoundException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Scanner;

public class day6_2 {
    public static void main(String[] args) {
        // use ../input/day6.test for using dummy input
        List<List<String>> firstBoard;
        String firstStep = "";
        int firstY = 0;
        int firstX = 0;
        try (Scanner scan = new Scanner(new File("../input/day6.input"))) {
            scan.useDelimiter("\n");

            firstBoard = new ArrayList<>();
            while (scan.hasNext()) {
                List<String> col = Arrays.asList(scan.next().trim().split(""));
                firstBoard.add(col);
            }
            for (int i = 0; i < firstBoard.size(); i++) {
                for (int j = 0; j < firstBoard.get(i).size(); j++) {
                    if ("^".equals(firstBoard.get(i).get(j)) || ">".equals(firstBoard.get(i).get(j))
                            || "v".equals(firstBoard.get(i).get(j)) || "<".equals(firstBoard.get(i).get(j))) {
                        firstY = i;
                        firstX = j;
                        firstStep = firstBoard.get(i).get(j);
                        break;
                    }
                }
            }

            boolean run = true;
            List<List<String>> board = new ArrayList<>();
            List<String> successList = new ArrayList<>();
            List<String> failedList = new ArrayList<>();
            // int counter = 0;
            while (run) {
                board = new ArrayList<>();
                for (List<String> s : firstBoard) {
                    List<String> abc = new ArrayList<>();
                    for (String a : s) {
                        abc.add(a);
                    }
                    board.add(abc);
                }
                int y = firstY;
                int x = firstX;
                String step = firstStep;
                boolean setStop = true;
                String currYX = "";
                while (setStop) {
                    if ("^".equals(step)) {
                        if (y == board.size() - 1) {
                            break;
                        } else if ("#".equals(board.get(y - 1).get(x))) {
                            x += 1;
                            step = ">";
                        } else {
                            y -= 1;
                        }
                    } else if (">".equals(step)) {
                        if (x == board.get(0).size() - 1) {
                            break;
                        } else if ("#".equals(board.get(y).get(x + 1))) {
                            y += 1;
                            step = "v";
                        } else {
                            x += 1;
                        }
                    } else if ("v".equals(step)) {
                        if (y == board.get(0).size() - 1) {
                            break;
                        } else if ("#".equals(board.get(y + 1).get(x))) {
                            x -= 1;
                            step = "<";
                        } else {
                            y += 1;
                        }
                    } else if ("<".equals(step)) {
                        if (x == 0) {
                            break;
                        } else if ("#".equals(board.get(y).get(x - 1))) {
                            y -= 1;
                            step = "^";
                        } else {
                            x -= 1;
                        }
                    }
                    if (!(successList.contains(y + "_" + x) || failedList.contains(y + "_" + x))) {
                        board.get(y).set(x, "O");
                        setStop = false;
                        currYX = y + "_" + x;
                    }
                }

                if (setStop) {
                    break;
                }

                step = firstStep;
                y = firstY;
                x = firstX;

                boolean runGame = true;
                boolean success = false;
                List<String> turn = new ArrayList<>();
                while (runGame) {
                    if ("^".equals(step)) {
                        if (y == 0) {
                            runGame = false;
                            success = false;
                        } else if ("#".equals(board.get(y - 1).get(x)) || "O".equals(board.get(y - 1).get(x))) {
                            board.get(y).set(x, "+");
                            String o = (y - 1) + "_" + (x) + step;
                            if (turn.contains(o)) {
                                runGame = false;
                                success = true;
                            } else {
                                turn.add(o);
                            }
                            step = ">";
                        } else {
                            if ("-".equals(board.get(y).get(x)) || "#".equals(board.get(y).get(x - 1))
                                    || "O".equals(board.get(y).get(x - 1))) {
                                board.get(y).set(x, "+");
                            } else {
                                board.get(y).set(x, "|");
                            }

                            y -= 1;
                        }
                    } else if (">".equals(step)) {
                        if (x == board.get(0).size() - 1) {
                            runGame = false;
                            success = false;
                        } else if ("#".equals(board.get(y).get(x + 1)) || "O".equals(board.get(y).get(x + 1))) {
                            board.get(y).set(x, "+");
                            String o = (y) + "_" + (x + 1) + step;
                            if (turn.contains(o)) {
                                runGame = false;
                                success = true;
                            } else {
                                turn.add(o);
                            }
                            step = "v";
                        } else {
                            if ("|".equals(board.get(y).get(x)) || "#".equals(board.get(y - 1).get(x))
                                    || "O".equals(board.get(y - 1).get(x))) {
                                board.get(y).set(x, "+");
                            } else {
                                board.get(y).set(x, "-");
                            }

                            x += 1;
                        }
                    } else if ("v".equals(step)) {
                        if (y == board.size() - 1) {
                            runGame = false;
                            success = false;
                        } else if ("#".equals(board.get(y + 1).get(x)) || "O".equals(board.get(y + 1).get(x))) {
                            board.get(y).set(x, "+");
                            String o = (y + 1) + "_" + (x) + step;
                            if (turn.contains(o)) {
                                runGame = false;
                                success = true;
                            } else {
                                turn.add(o);
                            }
                            step = "<";
                        } else {
                            if ("-".equals(board.get(y).get(x)) || "#".equals(board.get(y).get(x + 1))
                                    || "O".equals(board.get(y).get(x + 1))) {
                                board.get(y).set(x, "+");
                            } else {
                                board.get(y).set(x, "|");
                            }

                            y += 1;
                        }
                    } else if ("<".equals(step)) {
                        if (x == 0) {
                            runGame = false;
                            success = false;
                        } else if ("#".equals(board.get(y).get(x - 1)) || "O".equals(board.get(y).get(x - 1))) {
                            board.get(y).set(x, "+");
                            String o = (y) + "_" + (x - 1) + step;
                            if (turn.contains(o)) {
                                runGame = false;
                                success = true;
                            } else {
                                turn.add(o);
                            }
                            step = "^";
                        } else {
                            if ("|".equals(board.get(y).get(x)) || "#".equals(board.get(y + 1).get(x))
                                    || "O".equals(board.get(y + 1).get(x))) {
                                board.get(y).set(x, "+");
                            } else {
                                board.get(y).set(x, "-");
                            }

                            x -= 1;
                        }
                    }
                }

                if (success) {
                    successList.add(currYX);
                } else {
                    failedList.add(currYX);
                }

                // System.out.println("successList = " + successList);
                // System.out.println("failedList = " + failedList);
                // System.out.println("counter = " + counter);
                // counter++;

            }

            System.out.println(successList.size());
        } catch (FileNotFoundException e) {
            System.exit(1);
        }
    }
}

File: 2024/rust/src/days/day6_2.rs
use std::fs;

//slower then java, need optimizing
pub fn day6_2() {
    //let file_content = fs::read_to_string("../input/day6.test").expect("no");
    let file_content = fs::read_to_string("../input/day6.input").expect("no");
    println!("");

    let rows: Vec<&str> = file_content.trim().split('\n').collect();

    let (first_board, first_y, first_x, first_step) = get_board_y_x(&rows);

    let mut success_list: Vec<String> = Vec::new();
    let mut failed_list: Vec<String> = Vec::new();

    let mut counter = 0;
    'run: loop {
        let (board, curr_y_x, set_break) = set_block(
            &first_board,
            first_y,
            first_x,
            &first_step,
            &success_list,
            &failed_list,
        );

        if set_break {
            break 'run;
        }

        let success = run_game(&board, first_y, first_x, &first_step);
        if success {
            success_list.push(curr_y_x);
        } else {
            failed_list.push(curr_y_x);
        }

        //println!("{:?}", success_list);
        //println!("{:?}", failed_list);
        println!("counter = {}", counter);
        counter += 1;
    }
    println!("{}", success_list.len());
}
//
fn get_board_y_x(rows: &Vec<&str>) -> (Vec<Vec<String>>, usize, usize, String) {
    let mut board: Vec<Vec<String>> = Vec::new();
    let mut x = 0;
    let mut y = 0;
    let mut step = String::from("");

    let mut idy = 0;
    for row in rows {
        let cols: Vec<String> = row.trim().chars().map(|c| c.to_string()).collect();
        board.push(cols.clone());

        let mut idx = 0;
        if step == "" {
            for col in cols {
                if col == "^" || col == "v" || col == "<" || col == ">" {
                    x = idx;
                    y = idy;
                    step = col;
                }
                idx += 1;
            }
        }

        idy += 1;
    }

    (board, y, x, step)
}
fn set_block(
    first_board: &Vec<Vec<String>>,
    first_y: usize,
    first_x: usize,
    first_step: &String,
    success_list: &Vec<String>,
    failed_list: &Vec<String>,
) -> (Vec<Vec<String>>, String, bool) {
    let mut board = first_board.to_vec();
    let mut y = first_y;
    let mut x = first_x;
    let mut step = first_step.clone();
    let mut curr_y_x = String::from("");
    let mut set_break = true;
    loop {
        match &*step {
            "^" => {
                if y == 0 {
                    break;
                } else if board[y - 1][x] == "#" {
                    x += 1;
                    step = ">".to_string();
                } else {
                    y -= 1;
                }
            }
            ">" => {
                if x == board[0].len() - 1 {
                    break;
                } else if board[y][x + 1] == "#" {
                    y += 1;
                    step = "v".to_string();
                } else {
                    x += 1;
                }
            }
            "v" => {
                if y == board.len() - 1 {
                    break;
                } else if board[y + 1][x] == "#" {
                    x -= 1;
                    step = "<".to_string();
                } else {
                    y += 1;
                }
            }
            "<" => {
                if x == 0 {
                    break;
                } else if board[y][x - 1] == "#" {
                    y -= 1;
                    step = "^".to_string();
                } else {
                    x -= 1;
                }
            }
            _ => break,
        }
        curr_y_x = format!("{}_{}", y, x);
        if !(success_list.contains(&curr_y_x) || failed_list.contains(&curr_y_x)) {
            board[y][x] = "O".to_string();
            set_break = false;
            break;
        }
    }
    (board.to_vec(), curr_y_x, set_break)
}

fn run_game(board: &Vec<Vec<String>>, y: usize, x: usize, step: &String) -> bool {
    let mut board = board.to_vec();
    let mut y = y;
    let mut x = x;
    let mut step = step.clone();
    let mut turn: Vec<String> = Vec::new();
    let mut success = false;
    loop {
        match &*step {
            "^" => {
                if y == 0 {
                    break;
                } else if board[y - 1][x] == "#" || board[y - 1][x] == "O" {
                    board[y][x] = "+".to_string();
                    let t = format!("{}_{}{}", y - 1, x, step);
                    if turn.contains(&t) {
                        success = true;
                        break;
                    } else {
                        turn.push(t);
                    }
                    step = ">".to_string();
                } else {
                    if board[y][x] == "-" || board[y - 1][x] == "#" || board[y - 1][x] == "O" {
                        board[y][x] = "+".to_string();
                    } else {
                        board[y][x] = "|".to_string();
                    }
                    y -= 1;
                }
            }
            ">" => {
                if x == board[0].len() - 1 {
                    break;
                } else if board[y][x + 1] == "#" || board[y][x + 1] == "O" {
                    board[y][x] = "+".to_string();
                    let t = format!("{}_{}{}", y, x + 1, step);
                    if turn.contains(&t) {
                        success = true;
                        break;
                    } else {
                        turn.push(t);
                    }
                    step = "v".to_string();
                } else {
                    if board[y][x] == "|" || board[y][x + 1] == "#" || board[y][x + 1] == "O" {
                        board[y][x] = "+".to_string();
                    } else {
                        board[y][x] = "-".to_string();
                    }
                    x += 1;
                }
            }
            "v" => {
                if y == board.len() - 1 {
                    break;
                } else if board[y + 1][x] == "#" || board[y + 1][x] == "O" {
                    board[y][x] = "+".to_string();
                    let t = format!("{}_{}{}", y + 1, x, step);
                    if turn.contains(&t) {
                        success = true;
                        break;
                    } else {
                        turn.push(t);
                    }
                    step = "<".to_string();
                } else {
                    if board[y][x] == "-" || board[y + 1][x] == "#" || board[y + 1][x] == "O" {
                        board[y][x] = "+".to_string();
                    } else {
                        board[y][x] = "|".to_string();
                    }
                    y += 1;
                }
            }
            "<" => {
                if x == 0 {
                    break;
                } else if board[y][x - 1] == "#" || board[y][x - 1] == "O" {
                    board[y][x] = "+".to_string();
                    let t = format!("{}_{}{}", y, x - 1, step);
                    if turn.contains(&t) {
                        success = true;
                        break;
                    } else {
                        turn.push(t);
                    }
                    step = "^".to_string();
                } else {
                    if board[y][x] == "|" || board[y][x - 1] == "#" || board[y][x - 1] == "O" {
                        board[y][x] = "+".to_string();
                    } else {
                        board[y][x] = "-".to_string();
                    }
                    x -= 1;
                }
            }
            _ => break,
        }
    }
    success
}
